presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayElementBlock.java

```diff
@@ -0,0 +1,211 @@
+    {
+        checkReadablePosition(position);
+        return getBlock().bytesCompare(position + start, offset, length, otherSlice, otherOffset, otherLength);
+    }
+
+    @Override
+    public void writeBytesTo(int position, int offset, int length, BlockBuilder blockBuilder)
+    {
+        checkReadablePosition(position);
+        getBlock().writeBytesTo(position + start, offset, length, blockBuilder);
+    }
+
+    @Override
+    public void writePositionTo(int position, BlockBuilder blockBuilder)
+    {
+        checkReadablePosition(position);
+        getBlock().writePositionTo(position + start, blockBuilder);
+    }
+
+    @Override
+    public boolean equals(int position, int offset, Block otherBlock, int otherPosition, int otherOffset, int length)
+    {
+        checkReadablePosition(position);
+        return getBlock().equals(position + start, offset, otherBlock, otherPosition, otherOffset, length);
+    }
+
+    @Override
+    public int hash(int position, int offset, int length)
+    {
+        checkReadablePosition(position);
+        return getBlock().hash(position + start, offset, length);
+    }
+
+    @Override
+    public int compareTo(int leftPosition, int leftOffset, int leftLength, Block rightBlock, int rightPosition, int rightOffset, int rightLength)
+    {
+        checkReadablePosition(leftPosition);
+        return getBlock().compareTo(leftPosition + start, leftOffset, leftLength, rightBlock, rightPosition, rightOffset, rightLength);
+    }
+
+    @Override
+    public Block getSingleValueBlock(int position)
+    {
+        return getRegion(position, 1, true);
+    }
+
+    @Override
+    public boolean isNull(int position)
+    {
+        checkReadablePosition(position);
+        return getBlock().isNull(position + start);
+    }
+
+    @Override
+    public void assureLoaded()
+    {
+    }
+
+    @Override
+    public BlockEncoding getEncoding()
+    {
+        // ArrayElementBlock does not exist
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Block getRegion(int position, int length)
+    {
+        return getRegion(position, length, false);
+    }
+
+    @Override
+    public Block copyRegion(int position, int length)
+    {
+        return getRegion(position, length, true);
+    }
+
+    private Block getRegion(int position, int length, boolean compact)
+    {
+        // Suppose `ArrayElementBlock` exists. Although it can only represent 1 row. Implementation of `copyRegion` is different from `getRegion` because `ArrayElementBlock`
+        // is usually a one-item view into multiple items. It retains a much larger chunk of memory than necessary.
+
+        // ArrayElementBlock does not exist
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int getSizeInBytes()
+    {
+        return 2 * Integer.BYTES + getBlock().getSizeInBytes();
```

Comment:
It seems like this should be `getBlock().getRegion(start, getPositionCount()).getSizeInBytes()`, but that might be expensive, so I would move this to the subclass